Hang writing graphs during writing documentation

I have a big privat project, pretty big 900 or so functions builds fine in version 5 but on version 6 seems to get stuck in 
Writing resulting documentation.

a control c produces
```
Traceback (most recent call last):
  File "/opt/ford/6.0.0/bin/ford", line 11, in <module>
    load_entry_point('FORD==6.0.0', 'console_scripts', 'ford')()
  File "/opt/ford/6.0.0/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/__init__.py", line 391, in run
    main(proj_data,proj_docs,md)
  File "/opt/ford/6.0.0/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/__init__.py", line 379, in main
    docs.writeout()
  File "/opt/ford/6.0.0/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/output.py", line 192, in writeout
    if self.data['graph'].lower() == 'true': self.graphs.output_graphs(self.njobs)
  File "/opt/ford/6.0.0/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/graphmanager.py", line 174, in output_graphs
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
KeyboardInterrupt
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
KeyboardInterrupt
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
    results = pool.map(outputFuncWrap,args,len(args)/np)
KeyboardInterrupt
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 266, in map
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
KeyboardInterrupt
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
Traceback (most recent call last):
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 638, in get
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
    self.wait(timeout)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 635, in wait
    self._event.wait(timeout)
  File "/usr/lib64/python3.6/threading.py", line 551, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/lib64/python3.6/threading.py", line 295, in wait
    waiter.acquire()
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 335, in get
    res = self._reader.recv_bytes()
  File "/usr/lib64/python3.6/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/usr/lib64/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/usr/lib64/python3.6/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
KeyboardInterrupt
```

sorry this went under the radar.
here is the trace with --debug flag, does not look very different.
Also there is no error occurring it simply gets stuck... the trace i got but ctrl+c after waiting quite some time. I can give you access to the project if you want, just pm me, the project is hosted on gitlab.

```
Writing resulting documentation.
^CProcess ForkPoolWorker-7:
Process ForkPoolWorker-3:
Process ForkPoolWorker-5:
Process ForkPoolWorker-6:
Process ForkPoolWorker-8:
Process ForkPoolWorker-1:
Process ForkPoolWorker-2:
Traceback (most recent call last):
  File "/opt/ford/git/bin/ford", line 11, in <module>
Process ForkPoolWorker-4:
    load_entry_point('FORD==6.0.0', 'console_scripts', 'ford')()
  File "/opt/ford/git/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/__init__.py", line 391, in run
    main(proj_data,proj_docs,md)
  File "/opt/ford/git/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/__init__.py", line 379, in main
    docs.writeout()
  File "/opt/ford/git/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/output.py", line 192, in writeout
    if self.data['graph'].lower() == 'true': self.graphs.output_graphs(self.njobs)
  File "/opt/ford/git/lib/python3.6/site-packages/FORD-6.0.0-py3.6.egg/ford/graphmanager.py", line 175, in output_graphs
    results = pool.map(outputFuncWrap,args,len(args)/np)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 266, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 638, in get
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
KeyboardInterrupt
KeyboardInterrupt
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
    self.wait(timeout)
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 635, in wait
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
KeyboardInterrupt
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 335, in get
    res = self._reader.recv_bytes()
    self._event.wait(timeout)
  File "/usr/lib64/python3.6/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/usr/lib64/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/usr/lib64/python3.6/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib64/python3.6/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib64/python3.6/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
  File "/usr/lib64/python3.6/threading.py", line 551, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/lib64/python3.6/threading.py", line 295, in wait
    waiter.acquire()
KeyboardInterrupt
make[3]: *** [CMakeFiles/ford.dir/build.make:57: CMakeFiles/ford] Error 1
make[2]: *** [CMakeFiles/Makefile2:73: CMakeFiles/ford.dir/all] Interrupt
make[1]: *** [CMakeFiles/Makefile2:80: CMakeFiles/ford.dir/rule] In
```

--- a/ford/graphmanager.py
+++ b/ford/graphmanager.py
@@ -25,9 +25,9 @@
 import os
 import pathlib
 import sys
-from multiprocessing import Pool
 
 from tqdm import tqdm
+from tqdm.contrib.concurrent import process_map
 
 from ford.sourceform import (
     FortranFunction,
@@ -221,11 +221,12 @@ class GraphManager(object):
             )
             args.extend([(m.usesgraph, self.graphdir) for m in self.blockdata])
 
-            np = min(njobs, len(args))
-            pool = Pool(processes=np)
-            results = pool.map(outputFuncWrap, args, len(args) / np)  # noqa F841
-            pool.close()
-            pool.join()
+            process_map(
+                outputFuncWrap,
+                args,
+                max_workers=njobs,
+                desc="Writing graphs",
+            )
 
         if self.usegraph:
             self.usegraph.create_svg(self.graphdir)

--- a/ford/graphs.py
+++ b/ford/graphs.py
@@ -260,7 +260,8 @@ class BaseNode(object):
         return self.ident == other.ident
 
     def __hash__(self):
-        return hash(self.ident)
+        """Hash required to insert into dict"""
+        return id(self)
 
 
 class ModNode(BaseNode):

